src/lints/struct_with_no_pub_fields_changed.ron

```diff
@@ -0,0 +1,77 @@
+SemverQuery(
+    id: "struct_with_no_pub_fields_changed",
+    human_readable_name: "public API struct with no public fields is no longer a struct",
+    description: "A struct without pub fields was converted into an enum or union, breaking pattern matching.",
+    required_update: Major,
+    lint_level: Deny,
+    reference_link: Some("https://internals.rust-lang.org/t/rest-patterns-foo-should-match-non-struct-types/21607"),
+    reference: Some(
+        r#"\
+Even if a structs does not expose pub fields, pattern matching like `matches!(value, Example { .. })` is accessible outside their defining crate. \
+Changing such a struct to make it an enum or union will break this pattern match, as their syntax is different.
+
+More info: https://github.com/obi1kenobi/cargo-semver-checks/issues/954
+"#
+    ),
+    query: r#"
+    {
+        CrateDiff {
+            baseline {
+                item {
+                    ... on Struct {
+                        struct_typename: __typename @tag @output
+                        visibility_limit @filter(op: "=", value: ["$public"]) @output
+                        struct_type @output
+
+                        # Ensure the struct does not have pub fields to prevent overlap with struct_with_pub_fields_changed
+                        field @fold @transform(op: "count") @filter(op: "=", value: ["$zero"]) {
+                            visibility_limit @filter(op: "=", value: ["$public"])
+                            public_api_eligible @filter(op: "=", value: ["$true"])
```

Comment:
`public_api_eligible` is giving me a bit of (understandable) trouble understanding.

Do you think that https://predr.ag/blog/checking-semver-for-doc-hidden-items/ should be linked in the contribution guide?

On that note:
Is this correctly handled here?
My feeling is that `public_api_eligible` would need to be
- handled for the struct,
- the pub fields (hiding => not pub)
- and the the non-pub fields (hiding => not pub)
  
The last one is giving me a bit of grief, as the filters seem to be `and`-ed. I need it to be 
- `visibility_limit @filter(op: "!=", value: ["$public"])`
- OR `public_api_eligible @filter(op: "!=", value: ["$true"])`

Is there some docs or other lints which are using filters in this pattern?

Reply:
> `public_api_eligible` is giving me a bit of (understandable) trouble understanding.
> 
> Do you think that https://predr.ag/blog/checking-semver-for-doc-hidden-items/ should be linked in the contribution guide?

Possibly. It's _usually_ not that relevant to most lints (this one is an exception) so I wouldn't make it a "must read". But if you can find a good way to tuck in in as "extra reading" or something, that'd be good. Open a PR?

> On that note: Is this correctly handled here? My feeling is that `public_api_eligible` would need to be
> 
> * handled for the struct,
> * the pub fields (hiding => not pub)
> * and the the non-pub fields (hiding => not pub)
> 
> The last one is giving me a bit of grief, as the filters seem to be `and`-ed. I need it to be
> 
> * `visibility_limit @filter(op: "!=", value: ["$public"])`
> * **OR `public_api_eligible @filter(op: "!=", value: ["$true"])`**
> 
> using OR there or adding the counts of both would allow to remove the false negative for the testcase `PubStructWithNonPubDocFieldChangedToUnion`.

I think this is equivalent to "has some fields, but has no public API fields" right?

`constructible_struct_changed_type` requires that the struct have no fields at all, and also that it not be `#[non_exhaustive]`. So any fields are sufficient to escape that lint.

Which raises a good point: we probably want to catch the case of `#[non_exhaustive]` structs with no fields with a separate lint, since that will escape this lint's implementation. Would you like to take care of that lint next?

> Is there some docs or other lints which are using filters in this pattern? How wrong is my intuition on this topic?

Your intuition is pretty good! We don't have "OR" operators or addition operators yet, though I have a WIP branch for that in Trustfall so hopefully I can get it merged by the end of the year.

Reply:
> I think this is equivalent to "has some fields, but has no public API fields" right?

Yes, that works, was too focused on my path => did not zoom out to see that.

Reply:
No worries. Practice makes perfect, and I have a lot of practice doing this :)